evm_trace/vmtrace.py

```diff
@@ -0,0 +1,202 @@
+from __future__ import annotations
+
+from typing import Any, Dict, Iterator, List, Optional, Type, Union
+
+from eth.vm.memory import Memory  # type: ignore
+from eth.vm.stack import Stack  # type: ignore
+from eth_utils import to_checksum_address
+from hexbytes import HexBytes
+
+try:
+    from msgspec import Struct  # type: ignore
+    from msgspec.json import Decoder  # type: ignore
+except ImportError as e:
+    raise ImportError("msgspec not found. install it with `pip install msgspec`") from e
+
+# opcodes grouped by the number of items they pop from the stack
+# fmt: off
+POP_OPCODES = {
+    1: ["EXTCODEHASH", "ISZERO", "NOT", "BALANCE", "CALLDATALOAD", "EXTCODESIZE", "BLOCKHASH", "POP", "MLOAD", "SLOAD", "JUMP", "SELFDESTRUCT"],  # noqa: E501
+    2: ["SHL", "SHR", "SAR", "REVERT", "ADD", "MUL", "SUB", "DIV", "SDIV", "MOD", "SMOD", "EXP", "SIGNEXTEND", "LT", "GT", "SLT", "SGT", "EQ", "AND", "XOR", "OR", "BYTE", "SHA3", "MSTORE", "MSTORE8", "SSTORE", "JUMPI", "RETURN"],  # noqa: E501
+    3: ["RETURNDATACOPY", "ADDMOD", "MULMOD", "CALLDATACOPY", "CODECOPY", "CREATE"],
+    4: ["CREATE2", "EXTCODECOPY"],
+    6: ["STATICCALL", "DELEGATECALL"],
+    7: ["CALL", "CALLCODE"]
+}
+# fmt: on
+POPCODES = {op: n for n, opcodes in POP_OPCODES.items() for op in opcodes}
+POPCODES.update({f"LOG{n}": n + 2 for n in range(0, 5)})
+POPCODES.update({f"SWAP{i}": i + 1 for i in range(1, 17)})
+POPCODES.update({f"DUP{i}": i for i in range(1, 17)})
+
+
```

Comment:
What value does this library have? Why not just use int in type hint

Reply:
a fake uint256 type is needed to handle deserialization, since we need to convert hexstr to int